tests/unit/s2n_async_offload_cb_test.c

```diff
@@ -0,0 +1,224 @@
+    for (int test_idx = 0; test_idx < s2n_array_len(test_cases); test_idx++) {
+        for (int version_idx = 0; version_idx < s2n_array_len(versions); version_idx++) {
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_NOT_NULL(config);
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+            EXPECT_SUCCESS(s2n_config_set_verification_ca_location(config, S2N_DEFAULT_TEST_CERT_CHAIN, NULL));
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, versions[version_idx]));
+            if (test_cases[test_idx].client_auth) {
+                EXPECT_SUCCESS(s2n_config_set_client_auth_type(config, S2N_CERT_AUTH_REQUIRED));
+            }
+
+            struct s2n_async_offload_cb_test data = {
+                .async_test = test_cases[test_idx].async_test,
+                .result = test_cases[test_idx].cb_return,
+            };
+            EXPECT_SUCCESS(s2n_config_set_async_offload_callback(config, test_cases[test_idx].allow_list,
+                    s2n_async_offload_test_callback, &data));
+
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+            EXPECT_SUCCESS(s2n_connection_set_blinding(client_conn, S2N_SELF_SERVICE_BLINDING));
+            EXPECT_SUCCESS(s2n_set_server_name(client_conn, "localhost"));
+
+            DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+            EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+            s2n_error expected_error = test_cases[test_idx].expected_error;
+            if (test_cases[test_idx].async_test) {
+                EXPECT_SUCCESS(s2n_test_handshake_async(server_conn, client_conn, &data));
+            } else if (expected_error == S2N_ERR_OK) {
+                EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+            } else {
+                EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn), expected_error);
+            }
+            EXPECT_EQUAL(data.invoked_count, test_cases[test_idx].cb_invoked);
+
+            if (s2n_is_tls13_fully_supported() && version_idx == 1) {
+                EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS13);
+            }
```

Comment:
You're now confirming you're testing TLS1.3, but not TLS1.2. The requirement of this test is that we test both protocols, because they exercise two different code paths. If we removed openssl-1.0.2 support in the future, we could accidentally always be negotiating TLS1.3 instead of accidentally always be negotiating TLS1.2. If something is important to your test, don't assume-- verify!

```suggestion
            if (s2n_is_tls13_fully_supported() && version_idx == 1) {
                EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS13);
            } else {
                EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS12);
            }
```